### PR TITLE
PostingsForMatchers race with creating new series

### PR DESCRIPTION
**Pull request by dimitarvdimitrov**
### Bug

This bug can manifest when a new series is appended to the head block while PostingsForMatchers is computing its result. Then the result from PostingsForMatchers can contains series references that __do not match__ the matchers.

In particular, the posting lists which are used for subtraction (e.g. for `m != "0"`) can be computed before the other postings lists (e.g. `__name__ = "metric"`). 

This affects index queries (`/series`, LabelValues, LabelNames) even when there is transaction isolation enabled in the TSDB head. The reason is that isolation is only used for chunks filtering, but index queries do not use chunk iterators. This also affects running the TSDB with disabled transactions isolation.

### Example 1

The matchers are 
```
{datacenter!="dc1", datacenter=~"dc.*"}
```

PostingsForMatchers computes its result like so
* find the series IDs for `datacenter!=dc1`; call this $D1$
* find the series IDs for `datacenter=~dc.*`; call this $D2$
* return $P = D2 - D1$

If there is an append with a new timeseries between computing $D1$ and $D2$, then $P$ can contain more series than is correct.

* find the series IDs for `datacenter!=dc1`; call this $D1$
* __in another request__ append a new timeseries to the TSDB head `{datacenter="dc1"}`
* find the series IDs for `datacenter=~dc.*`; call this $D2$
* return $P = D2 - D1$

There is no synchronisation between the calculation of the different series sets, so now $D2$ contains the new timeseries, but $D1$ doesn't.

### Example 2

With matchers 

```
{datacenter!="dc1"}
```

The process is very similar to the one above, but the value of $D2$ is the list of all postings.

### Tests

I added two test cases that exemplify the bug. They are different but are executed similarly by PostingsForMatchers:

* `__name__ = "metric", m != "0"` - instead of using all postings, this case uses the `m=0` posting list as in example 1
* `__name__ != "metric"` - involves getting the list of all postings as in example 2

Reviewers can checkout the test commit to see them fail without the fix.

### Fix

Always compute the series IDs for intersecting matchers first and for subtracting matchers second. This means that intersected sets of series will not contain new series. Instead subtracted sets will contain new series. When the subtracted sets contain new series, then the result is still valid (only slightly outdated).

### CODE REVIEW

File: tsdb/querier_test.go
Comment (colega) [inline tsdb/querier_test.go:2254]:
```suggestion
	for i := 0; ctx.Err() != nil; i++ {
```

File: tsdb/querier_test.go
Comment (colega) [inline tsdb/querier_test.go:2264]: I wonder if just sleeping 1ms between each series would have a better effect. Because what we really want is to add a series at the same time when we're calculating postings. Adding 1 or 10 doesn't make the test different, but sleeping between each append makes it more prone to switch to scheduling the other goroutine.

File: tsdb/querier_test.go
Comment (colega) [inline tsdb/querier_test.go:2240]: Nit: I would be more explicit on the label name, requesting `"this_label_does_not_exist"`, that would save some moments to the reader checking the test 3 times to see that this label isn't defined.

Comment (jesusvazquez) [reply tsdb/querier_test.go:2240]: Agreed on this comment, it wasn'tt straightforward for me to know what was happening here.

File: tsdb/querier.go
Comment (jesusvazquez) [inline tsdb/querier.go:254]: Nit: I prefer `hasSubstractingMatchers, hasIntersectingMatchers` but its fine if you don't want to change this

**Review by jesusvazquez — COMMENTED** [review]
Nice find. Overall both code tests and comments are looking great. Waiting on some of the feedback that @colega left you in case you want to address it. Will make a second pass later today and merge this if it has been addressed.

File: tsdb/querier_test.go
Comment (dimitarvdimitrov) [reply tsdb/querier_test.go]: In this case `n` always exists, but we expect to get, no values because the matchers select no series. I changed `m` to `always_0` and `n` to `seq` (from "sequence") in c52f29ccf324a11b2d02368105d5e4840e59bedd. Does this make it better?

File: tsdb/querier_test.go
Comment (dimitarvdimitrov) [reply tsdb/querier_test.go]: good call, with this change (sleep 1ms, every append) I'm able to reproduce the original bug slightly more often than with the previous setup (sleep 10ms, every 10 appends). Made the change in 5457f68e5

**Comment by dimitarvdimitrov** [comment]
bump, @jesusvazquez  :)

**Comment by jesusvazquez** [comment]
Looks good. To be extra safe I'm going to start prombench and compare with the latest release to discard regressions on the query path for the next maintainer that crafts a release.

**Comment by bwplotka** [comment]
See -team email thread for details. Restarting.

/prombench cancel

**Comment by jesusvazquez** [comment]
Benchmark has been running since yesterday and results are finally up.

First 30 minutes showed up very small differences in CPU and memory allocations but over time these have gone so we can obviate them.

image
image

**Review by jesusvazquez — APPROVED** [review]

**Pull request merged**
